presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java

```diff
@@ -349,4 +408,179 @@ else if (isCovariantParameterPosition(commonSuperTypeBase.get(), i)) {
 
         return Optional.of(new TypeSignature(commonSuperTypeBase.get(), typeParameters.build()));
     }
+
+    private static Optional<TypeSignature> getCommonSuperTypeForLiteralTypes(TypeSignaturePair typeSignaturePair)
+    {
+        checkState(typeSignaturePair.containsAnyOf(StandardTypes.VARCHAR, StandardTypes.DECIMAL));
+
+        for (String varcharSubType : new String[] {RegexpType.NAME, LikePatternType.NAME, JsonPathType.NAME}) {
+            if (typeSignaturePair.is(StandardTypes.VARCHAR, varcharSubType)) {
+                return Optional.of(typeSignaturePair.get(varcharSubType));
+            }
+        }
+
+        if (typeSignaturePair.is(StandardTypes.VARCHAR, UnknownType.NAME)) {
+            return getCommonSuperTypeForVarchar(new TypeSignaturePair(
+                    typeSignaturePair.get(StandardTypes.VARCHAR),
+                    new TypeSignature(VARCHAR, ImmutableList.of(TypeSignatureParameter.of(0)))
+            ));
+        }
+
+        if (typeSignaturePair.is(StandardTypes.VARCHAR, StandardTypes.VARCHAR)) {
+            return getCommonSuperTypeForVarchar(typeSignaturePair);
+        }
+
+        if (typeSignaturePair.is(StandardTypes.DECIMAL, UnknownType.NAME)) {
+            return getCommonSuperTypeForDecimals(new TypeSignaturePair(
+                    typeSignaturePair.get(StandardTypes.DECIMAL),
+                    createDecimalTypeSignature(1, 0)
+            ));
+        }
+
```

Comment:
some comment what it means. Or generally describe logic of this method in 2-3 lines of comment at the beginning.